### PR TITLE
Implement implement qwxwayland signals from wlr_xwayland

### PR DESCRIPTION
**Pull request by justforlxz**

### CODE REVIEW

File: src/types/qwxwayland.cpp [hunk 1]
```diff
@@ -18,20 +20,69 @@ extern "C" {
...
+    QWXWaylandPrivate(wlr_xwayland *handle, bool isOwner, QWXWayland *qq)
+        : QWObjectPrivate(handle, isOwner, qq)
+    {
+        Q_ASSERT(!map.contains(handle));
+        map.insert(handle, qq);
+        sc.connect(&handle->events.ready, this, &QWXWaylandPrivate::on_ready);
+        sc.connect(&handle->events.new_surface, this, &QWXWaylandPrivate::on_new_surface);
+        sc.connect(&handle->events.remove_startup_info, this, &QWXWaylandPrivate::on_remove_stratup_info);
+    }
+    ~QWXWaylandPrivate() {
```
Comment (zccrs) [inline src/types/qwxwayland.cpp:35]: See https://github.com/vioken/qwlroots/blob/master/src/class_template.txt#L59
Keep template code.

File: src/types/qwxwayland.cpp [hunk 2]
```diff
@@ -18,20 +20,69 @@ extern "C" {
...
+    QW_DECLARE_PUBLIC(QWXWayland)
+    QWSignalConnector sc;
+};
+QHash<void*, QWXWayland*> QWXWaylandPrivate::map;
+
 QWXWayland *QWXWayland::create(QWDisplay *wl_display, QWCompositor *compositor, bool lazy)
 {
-    auto* pointer = wlr_xwayland_create(wl_display->handle(), compositor->handle(), lazy);
-    return pointer ? from(pointer) : nullptr;
+    return new QWXWayland(wlr_xwayland_create(wl_display->handle(), compositor->handle(), lazy), true);
```
Comment (zccrs) [inline src/types/qwxwayland.cpp:72]: The return value of `wlr_xwayland_create` maybe is nullptr

File: src/types/qwxwayland.cpp [hunk 3]
```diff
@@ -54,7 +105,13 @@ void QWXWayland::setSeat(QWSeat *seat)
 
 wlr_xwayland* QWXWayland::handle() const
 {
-    return reinterpret_cast<wlr_xwayland*>(const_cast<QWXWayland*>(this));
+    return QWObject::handle<wlr_xwayland>();
+}
+
+QWXWayland::QWXWayland(wlr_xwayland* handle, bool isOwner)
+    : QObject()
```
Comment (zccrs) [inline src/types/qwxwayland.cpp:112]: Where is parent argument?

File: src/types/qwxwayland.h [hunk 4]
```diff
@@ -14,19 +15,28 @@ QW_BEGIN_NAMESPACE
...
+    static QWXWayland* create(QWDisplay *wl_display, QWCompositor *compositor, bool lazy);
+    static QWXWayland* from(wlr_xwayland* handle);
+    static QWXWayland* get(wlr_xwayland* handle);
     wlr_xwayland* handle() const;
 
-    void destroy();
     void setCursor(const QImage& image, const QPoint& hotspot);
     void setSeat(QWSeat *seat);
+
+Q_SIGNALS:
```
Comment (zccrs) [inline src/types/qwxwayland.h:32]: The signals's name keep Qt style.

**Review by zccrs — CHANGES_REQUESTED** [review]

File: src/types/qwxwayland.cpp [hunk 5]
```diff
@@ -18,20 +20,78 @@ extern "C" {
...
+        Q_ASSERT(m_handle);
+        Q_ASSERT(map.contains(m_handle));
+        Q_EMIT q_func()->beforeDestroy(q_func());
+        map.remove(m_handle);
+        sc.invalidate();
+    }
+
+    void on_ready() {
+        Q_ASSERT(m_handle);
+        Q_ASSERT(map.contains(m_handle));
```
Comment (wineee) [inline src/types/qwxwayland.cpp:56]: 应该不是每个信号都要检查这个， destroy有应该就够了

Comment (justforlxz) [reply src/types/qwxwayland.cpp]: 反正是无参的，直接绑定到public class上面了。

File: src/types/qwxwayland.h [hunk 6]
```diff
@@ -14,19 +15,29 @@ QW_BEGIN_NAMESPACE
 class QWDisplay;
 class QWCompositor;
 class QWSeat;
-class QWXWayland
+class QWXWaylandPrivate;
+class QWXWayland : public QObject, public QWObject
```
Comment (zccrs) [inline src/types/qwxwayland.h:19]: `QW_EXPORT`

**Review by zccrs — CHANGES_REQUESTED** [review]

**Review by zccrs — APPROVED** [review]

**Pull request merged**